tests/order_with_expression/tests.py

```diff
@@ -0,0 +1,131 @@
+from operator import attrgetter
+
+from django.test import TestCase
+from django.db import models
+
+from .models import Musician, Album, Song
+
+
+class OrderWithExpressionTests(TestCase):
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.musician1 = Musician.objects.create(instrument='mayonaise')
+        cls.musician3 = Musician.objects.create(instrument=None)
+        cls.musician2 = Musician.objects.create(instrument='zoo')
+
+        cls.album1 = Album.objects.create(
+            artist=cls.musician1,
+            name='Album 1',
+        )
+        cls.album3 = Album.objects.create(
+            artist=cls.musician3,
+            name='Album 3',
+        )
+        cls.album2 = Album.objects.create(
+            artist=cls.musician2,
+            name='Album 2',
+        )
+
+        cls.song1 = Song.objects.create(
+            album=cls.album1,
+            name='Song 1',
+        )
+        cls.song3 = Song.objects.create(
+            album=cls.album3,
+            name='Song 3',
+        )
+        cls.song2 = Song.objects.create(
+            album=cls.album2,
+            name='Song 2',
+        )
+
+    def test_order_by_expression_nulls_last(self):
+        self.assertQuerysetEqual(
+            Musician.objects.all(),
+            ['mayonaise', 'zoo', None],
+            attrgetter('instrument')
+        )
+
+    def test_order_by_related_expression_nulls_last(self):
+        self.assertQuerysetEqual(
+            Album.objects.all(),
+            ['Album 1', 'Album 2', 'Album 3'],
+            attrgetter('name')
+        )
+
+    def test_order_by_related_expression_through_related_field_nulls_last(self):
+        self.assertQuerysetEqual(
+            Song.objects.all(),
+            ['Song 1', 'Song 2', 'Song 3'],
+            attrgetter('name')
+        )
+
+    def test_order_by_expression_nulls_first(self):
+
+        ordering = [models.F('instrument').asc(nulls_first=True)]
+
+        with mock.patch.object(Musician._meta, 'ordering', ordering):
+            self.assertQuerysetEqual(
+                Musician.objects.all(),
+                [None, 'mayonaise', 'zoo'],
+                attrgetter('instrument')
+            )
+
+    def test_order_by_related_expression_nulls_first(self):
+
+        ordering = [models.F('instrument').asc(nulls_first=True)]
+
+        with mock.patch.object(Musician._meta, 'ordering', ordering):
+            self.assertQuerysetEqual(
+                Album.objects.all(),
+                ['Album 3', 'Album 1', 'Album 2'],
+                attrgetter('name')
+            )
+
+    def test_depth_3_nulls_first(self):
+
+        ordering = [models.F('instrument').asc(nulls_first=True)]
+
+        with mock.patch.object(Musician._meta, 'ordering', ordering):
+            self.assertQuerysetEqual(
+                Song.objects.all(),
+                ['Song 3', 'Song 1', 'Song 2'],
+                attrgetter('name')
+            )
+
+    @skip('What should the behavior be?')
```

Comment:
Not sure what the expected behavior here is.

```python
class Album:
    class Meta:
        # Should both of these work the same way? Or does
        # The first mean "order by artist_id"
        # And the second mean, order by the default order of Musician?
        ordering = [models.F('artist').asc(nulls_last=True)]
        ordering = ['artist']
```

Reply:
`ordering = ['artist']` will mean order by artist (rather than the default order of the related model)

Reply:
@carltongibson no, you should use`artist_id` if you want to achieve this. [There's a documented difference between `order_by('artist')` and `order_by('artist_id')`](https://docs.djangoproject.com/en/2.1/ref/models/querysets/#order-by) as it was not even possible to disable related ordering before 24ec9538b7ca400f68ba08fab380445ca95d7c02.

Given `Artist.Meta.ordering = ['foo', F('bar').asc(nulls_last=False)]`

then `Album.Meta.ordering = [models.F('artist').asc(nulls_last=True)]`should be the equivalent of

`[F('artist__foo').asc(nulls_last=True), F('artist__bar').asc(nulls_last=True)]`

Reply:
Good. Thanks Simon.